build/source/engine/allocspace.f90

```diff
@@ -57,136 +56,194 @@ module allocspace_module
+ character(*),intent(out)             :: message            ! error message
+ ! define general variables
+ character(len=256)                   :: cmessage           ! error message for downwind routine
+ character(LEN=256)                   :: infile             ! input filename
+ 
+ ! define variables for NetCDF file operation
+ integer(i4b)                         :: mode               ! netCDF file open mode
+ integer(i4b)                         :: ncid               ! integer variables for NetCDF IDs
+ integer(i4b)                         :: varid              ! variable id from netcdf file
+ integer(i4b)                         :: hruDimID           ! integer variables for NetCDF IDs
+ integer(i4b)                         :: gruDimID           ! integer variables for NetCDF IDs
+ integer(i4b),allocatable             :: hru_id(:)          ! unique id of hru over entire domain
+ integer(i4b),allocatable             :: gru_id(:)          ! unique ids of GRUs
+ integer(i4b),allocatable             :: hru2gru_id(:)      ! unique GRU ids at each HRU
+      
+ ! define local variables
+ integer(i4b)                         :: hruCount           ! number of hrus in a gru
+ integer(i4b)                         :: iGRU               ! index of a GRU and HRU
+
+ ! Start procedure here
+ err=0; message="allocate_gru_hru_map/"
+
+ ! check that gru_struc structure is initialized
+ if(allocated(gru_struc)) deallocate(gru_struc)
+ if(allocated(index_map)) deallocate(index_map)
+
+ ! build filename
+ infile = trim(SETNGS_PATH)//trim(LOCAL_ATTRIBUTES)
+ ! open file
+ mode=nf90_noWrite
+ call nc_file_open(trim(infile), mode, ncid, err, cmessage)
+ if(err/=0)then; message=trim(message)//trim(cmessage); return; endif
+
+ ! get gru_ix dimension length
+ err = nf90_inq_dimid(ncid, "ngru", gruDimID);             if(err/=0)then; message=trim(message)//'problem finding nGRU dimension'; return; endif
+ err = nf90_inquire_dimension(ncid, gruDimID, len = nGRU); if(err/=0)then; message=trim(message)//'problem reading nGRU dimension'; return; endif
+
+ ! get hru_dim dimension length (ie., the number of hrus in entire domain)
+ err = nf90_inq_dimid(ncid, "nhru", hruDimID);             if(err/=0)then; message=trim(message)//'problem finding nHRU dimension'; return; endif
+ err = nf90_inquire_dimension(ncid, hruDimID, len = nHRU); if(err/=0)then; message=trim(message)//'problem reading nHRU dimension'; return; endif
+ 
+ allocate(gru_struc(nGRU), index_map(nHRU), stat=err)
+ if(err/=0)then; err=20; message=trim(message)//'problem allocating space for mapping structures'; return; endif
+ 
+ allocate(gru_id(nGRU),hru_id(nHRU),hru2gru_id(nHRU),stat=err)
+ if(err/=0)then; err=20; message=trim(message)//'problem allocating space for zLocalAttributes gru-hru correspondence vectors'; return; endif
+
+ ! read gru_id from netcdf file
+ err = nf90_inq_varid(ncid, "gruId", varid);     if(err/=0)then; message=trim(message)//'problem finding gruId variable'; return; endif
```

Comment:
I prefer to include error messages in the code, so we know what the error is without having to go and look in a separate table.